_posts/2025-03-11-sqlite4j-pure-java-sqlite.adoc

```diff
@@ -0,0 +1,127 @@
+
+Here's a sample implementation:
+
+[source,java]
+----
+@ApplicationScoped
+public class SQLiteBackup {
+    @ConfigProperty(name = "quarkus.datasource.jdbc.url")
+    String jdbcUrl;
+
+    @Inject
+    AgroalDataSource dataSource;
+
+    // Execute a backup every 10 seconds
+    @Scheduled(delay=10, delayUnit=TimeUnit.SECONDS, every="10s")
+    void scheduled() { backup(); }
+
+    // Execute a backup during shutdown
+    public void onShutdown(@Observes ShutdownEvent event) { backup(); }
+
+    void backup() {
+        String dbFile = jdbcUrl.substring("jdbc:sqlite:".length());
+
+        var originalDbFilePath = Paths.get(dbFile);
+        var backupDbFilePath = originalDbFilePath
+                                    .toAbsolutePath()
+                                    .getParent()
+                                    .resolve(originalDbFilePath.getFileName() + "_backup");
+
+        try (var conn = dataSource.getConnection();
+                var stmt = conn.createStatement()) {
+            // Execute the backup
+            stmt.executeUpdate("backup to " + backupDbFilePath);
+            // Atomically replace the DB file with its backup
+            Files.move(backupDbFilePath, originalDbFilePath,
+                StandardCopyOption.ATOMIC_MOVE,
+                StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException | SQLException e) {
+            throw new RuntimeException("Failed to back up the database", e);
+        }
+    }
+}
+----
+
+== Technical Deep Dive
+
+https://github.com/roastedroot/sqlite4j[sqlite4j] compiles the official SQLite C https://www.sqlite.org/amalgamation.html[amalgamation] to WebAssembly (Wasm), which is then translated into Java bytecode using the https://chicory.dev/docs/experimental/aot[Chicory AOT compiler].
+This enables SQLite to run in a pure Java environment while maintaining its full functionality.
+
+image::sqlite-compilation.png[width=80%, align="center", alt="SQLite compilation"]
+
+== Security and Isolation
+
+One of the key benefits of this approach is security.
+Running SQLite inside a Wasm sandbox ensures memory safety and isolates it from the host system, making it an excellent choice for applications that require embedded databases while avoiding the risks of native code execution.
+
+== Conclusion
+
+With the new https://github.com/quarkiverse/quarkus-jdbc-sqlite4j[quarkus-jdbc-sqlite4j] extension, you get the best of both worlds: the power and reliability of SQLite combined with the safety and portability of Java.
+This extension seamlessly integrates SQLite into Quarkus applications while maintaining a lightweight and secure architecture. Best of all, everything compiles effortlessly with __native-image__.
+
+Ready to try it out? Give https://github.com/quarkiverse/quarkus-jdbc-sqlite4j[quarkus-jdbc-sqlite4j] a spin in your projects and experience the benefits of running SQLite in pure Java with Quarkus!
```

Comment:
```suggestion
Ready to try it out? Give https://github.com/quarkiverse/quarkus-jdbc-sqlite4j[quarkus-jdbc-sqlite4j] a spin in your projects and experience the benefits of running SQLite in pure Java with Quarkus!

We are looking to gather interest and reports from production usecases!
```

Reply:
just to encourage feedback and people realize that this is a new experimental stack

Reply:
nice :+1: